src/m3ninx/index/segment/fst/encoding/docs/bench_test.go

```diff
@@ -0,0 +1,253 @@
+)
+
+func TestCompressionWriterAllocs(t *testing.T) {
+	w, err := newCompressedDataWriter(CompressionOptions{
+		Type:     DeflateCompressionType,
+		PageSize: 1 << 14,
+	}, nil)
+	require.NoError(t, err)
+	w.Reset(ioutil.Discard)
+	_, err = w.Write(testDocuments[0])
+	require.NoError(t, err)
+	require.NoError(t, w.Flush())
+
+	w.Reset(ioutil.Discard)
+	_, err = w.Write(testDocuments[0])
+	require.NoError(t, err)
+	require.NoError(t, w.Flush())
+}
+
+/*
+$ go test ./src/m3ninx/index/segment/fst/encoding/docs -run 'Benchmark' -bench=BenchmarkWriteCompressedStoredFieldsData -v
+goos: darwin
+goarch: amd64
+pkg: github.com/m3db/m3/src/m3ninx/index/segment/fst/encoding/docs
+BenchmarkWriteCompressedStoredFieldsData/snappy-4         	10000000	       185 ns/op	      11 B/op	       0 allocs/op
+BenchmarkWriteCompressedStoredFieldsData/deflate-4        	 2000000	       814 ns/op	       2 B/op	       0 allocs/op
+BenchmarkWriteCompressedStoredFieldsData/lz4-4            	 3000000	       369 ns/op	     554 B/op	       0 allocs/op
+PASS
+ok  	github.com/m3db/m3/src/m3ninx/index/segment/fst/encoding/docs	6.124s
+*/
+func BenchmarkWriteCompressedStoredFieldsData(b *testing.B) {
+	for _, codec := range []CompressionType{SnappyCompressionType, DeflateCompressionType, LZ4CompressionType} {
+		b.Run(codec.String(), func(b *testing.B) {
+			var buf bytes.Buffer
+			opts := CompressionOptions{
+				Type:     codec,
+				PageSize: 1 << 14,
+			}
+			w, err := newCompressedDataWriter(opts, nil)
+			if err != nil {
+				panic(err.Error())
+			}
+			w.Reset(&buf)
+			b.ResetTimer()
+			b.ReportAllocs()
+			for j := 0; j < b.N; j++ {
+				i := b.N % len(testDocuments)
+				d := testDocuments[i]
+				_, err := w.Write(d)
+				if err != nil {
+					panic(err.Error())
+				}
+			}
+			if err := w.Flush(); err != nil {
+				panic(err.Error())
+			}
+		})
+	}
+}
+
+/*
+$ go test ./src/m3ninx/index/segment/fst/encoding/docs -run 'Benchmark' -bench=BenchmarkAccessSingleCompressedStoredFieldsData -v
+goos: darwin
+goarch: amd64
+pkg: github.com/m3db/m3/src/m3ninx/index/segment/fst/encoding/docs
+BenchmarkAccessSingleCompressedStoredFieldsData/uncached_snappy-4         	  100000	     12433 ns/op	1317.78 MB/s	   16560 B/op	       4 allocs/op
+BenchmarkAccessSingleCompressedStoredFieldsData/uncached_deflate-4        	   10000	    161008 ns/op	 101.76 MB/s	   57904 B/op	      19 allocs/op
+BenchmarkAccessSingleCompressedStoredFieldsData/uncached_lz4-4            	   20000	     58620 ns/op	 279.49 MB/s	  188944 B/op	       9 allocs/op
+BenchmarkAccessSingleCompressedStoredFieldsData/cached_snappy-4           	  200000	      7031 ns/op	2330.03 MB/s	     128 B/op	       2 allocs/op
+BenchmarkAccessSingleCompressedStoredFieldsData/cached_deflate-4          	   10000	    137704 ns/op	 118.98 MB/s	     832 B/op	      11 allocs/op
+BenchmarkAccessSingleCompressedStoredFieldsData/cached_lz4-4              	  100000	     10149 ns/op	1614.26 MB/s	     129 B/op	       2 allocs/op
+PASS
```

Comment:
Looks like Snappy wins on speed and allocations with reads, especially for uncached?

Reply:
Yeah exactly. I'm inclined to use it for a v1, and subsequently spend time on optimizing the lz4 lib (it should be able to do a lot better than the current implementation) if we need better.